quixstreams/sinks/community/elasticsearch.py

```diff
@@ -0,0 +1,189 @@
+        :param on_client_connect_failure: An optional callback made after failed
+            client authentication (which should raise an Exception).
+            Callback should accept the raised Exception as an argument.
+            Callback must resolve (or propagate/re-raise) the Exception.
+        :param kwargs: additional kwargs that are passed to the ElasticSearch client
+        """
+        super().__init__(
+            on_client_connect_success=on_client_connect_success,
+            on_client_connect_failure=on_client_connect_failure,
+        )
+
+        self._client_kwargs = {"hosts": url, **kwargs}
+        self._url = url
+        self._index = index
+        self._batch_size = batch_size
+        self._max_bulk_retries = max_bulk_retries
+        self._ignore_bulk_errors = ignore_bulk_upload_errors
+        self._document_id_setter = document_id_setter
+        self._mapping = mapping or DEFAULT_MAPPING
+        self._add_message_metadata = add_message_metadata
+        self._add_topic_metadata = add_topic_metadata
+        self._client: Optional[Elasticsearch] = None
+
+    def setup(self):
+        self._client = Elasticsearch(**self._client_kwargs)
+        self._client.info()
+        self._create_index()
+
+    def _create_index(self):
+        """Create index if it doesn't exist."""
+        if not self._client.indices.exists(index=self._index):
+            logging.info(
+                f"Creating Index '{self._index}' created with mapping: '{self._mapping}'..."
+            )
+            self._client.indices.create(index=self._index, body=self._mapping)
+        else:
+            logging.debug(f"Index '{self._index}' already exists.")
+
+    def _add_metadata(self, topic: str, partition: int, record: SinkItem) -> dict:
+        value = record.value
+        if self._add_message_metadata:
+            # tuples and bytes are one of the few unsupported common types, so we
+            # convert them so users don't need to manage it.
+            key = record.key
+            if isinstance(key, tuple):
+                key = [k.decode() if isinstance(k, bytes) else k for k in key]
+            elif isinstance(key, bytes):
+                key = key.decode()
+            value["__key"] = key
+            value["__headers"] = {
+                k: v.decode() if isinstance(v, bytes) else v
+                for k, v in dict(record.headers or {}).items()
+            }
+            value["__timestamp"] = record.timestamp
+        if self._add_topic_metadata:
+            value["__topic"] = topic
+            value["__partition"] = partition
+            value["__offset"] = record.offset
+        return value
+
+    def write(self, batch: SinkBatch) -> None:
+        start = time.monotonic()
+        min_timestamp = sys.maxsize
+        max_timestamp = -1
+
+        if self._add_message_metadata or self._add_topic_metadata:
+            add_metadata = partial(self._add_metadata, batch.topic, batch.partition)
```

Comment:
could we make this a boolean and call `self._add_metadata` directly if it's true ?